Context menu

maier49 opened the pull request:
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [x] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [x] WidgetProperties are exported

<!--
Our bots should ensure:

* [ ] All contributors have signed a CLA
* [ ] The PR passes CI testing
* [ ] Code coverage is maintained
* [ ] The PR has been reviewed and approved
-->

**Description:**

Resolves #938

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-f6rya87dr.now.sh
* Tests Deployment: https://dojowidgets-ankjcdmjh.now.sh

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-md7ampjjc.now.sh
* Tests Deployment: https://dojowidgets-a66d6js13.now.sh

tomdye commented on src/context-menu/index.tsx:
```diff
@@ -0,0 +1,91 @@
+
+export interface ContextMenuProperties {
+	options: MenuOption[];
+	onValue(value: string): void;
```
As this is a menu with selection, do you think `onSelect` might be better?

maier49 replied:
Sure, I'll update that.

tomdye commented on src/context-menu/index.tsx:
```diff
@@ -0,0 +1,91 @@
+export interface ContextMenuProperties {
+	options: MenuOption[];
+	onValue(value: string): void;
+	menuWidth?: number;
```
Do we really need `menuWidth` here? The `popup` widget uses `dimensions` to figure out the size of the menu where appropriate.

maier49 replied:
I'm not sure using `dimensions` would make sense for a context menu. In the example the context menu can be applied to any region of text. In general you could be applying the context menu to any kind of content, and unlike the popup which sizes to the specific content it's attached to I would think that the context menu's width should be consistent regardless of what component is clicked on?

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-md7ampjjc.now.sh
* Tests Deployment: https://dojowidgets-a66d6js13.now.sh

tomdye commented on src/context-menu/index.tsx:
```diff
@@ -0,0 +1,91 @@
+					>
+						<Menu
+							key="menu"
+							theme={{
```
bear in mind this will need to change after https://github.com/dojo/widgets/pull/940

maier49 replied:
👍 I believe I've updated it accordingly

tomdye commented on src/examples/src/widgets/context-menu/Basic.tsx:
```diff
@@ -0,0 +1,42 @@
+			>
+				{text}
+			</ContextMenu>
+			<div>Printed Text: {printedText}</div>
```
what impact does this usage have on positioning / layout of the chid this is passed?

maier49 replied:
It wraps it in a single `div` with no additional styles, but I added a `contentWrapper` class to the default theme so that you could style the wrapper however you'd like.

maier49 commented on src/context-menu/index.tsx:
```diff
@@ -0,0 +1,91 @@
+				key="content"
+				oncontextmenu={(event: MouseEvent) => {
+					event.preventDefault();
+					icache.set('x', event.clientX - 2);
```
I'll update those to be constants, it's just an offset to make it look like the menu's coming off your cursor the same way a real context menu would, though obviously this wouldn't take into account a different sized cursor.

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-gpwzuy4vb.now.sh
* Tests Deployment: https://dojowidgets-ak6lcf5ck.now.sh

tomdye commented:
Looks like focus / mouse hover is still not working on the example

maier49 commented:
@tomdye 
I think I should have all feedback addressed now:
* Added check for whether the menu will fit under the cursor or needs to be placed above it (and initial transparent render until this is determined)
* Using the focus meta to focus the menu when its opened
* Added a border around the menu in the default theme following the example of the select
* Renamed `menuWrapper` class to `wrapper` to not conflict with the prefixed menu classes
* Updated positioning so it works if you're scrolled 🤦‍♂

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-7cmibeviw.now.sh
* Tests Deployment: https://dojowidgets-6qblvyjdn.now.sh

tomdye commented:
Thanks for continuing to work on this @maier49. It seems now that there is a lot of overlap between `Popup` and `ConextMenu` in terms of positioning and space calculation. Do you think it might be better than the further abstract `Popup` into `ContextPopup` / `TriggerPopup` which use a common `Popup` to do those calcs under the hood? That would remove the need for this code duplication and give us ways to popup _anything_ on demand via right click or via a trigger?

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-btmjjm7uo.now.sh
* Tests Deployment: https://dojowidgets-cvalttv43.now.sh

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-bj2ryebu3.now.sh
* Tests Deployment: https://dojowidgets-iqfd04aot.now.sh

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-mq7zqzhb7.now.sh
* Tests Deployment: https://dojowidgets-ls4ommrkx.now.sh

tomdye commented on src/context-menu/index.tsx:
```diff
@@ -0,0 +1,45 @@
+import ContextPopup from '../context-popup';
+
+export interface ContextMenuProperties {
+	options: MenuOption[];
```
can we get comments on these please?

tomdye commented on src/context-menu/context-menu.m.css:
```diff
@@ -0,0 +1,13 @@
+.root {
```
comments please

maier49 replied:
I've removed these CSS files, as you mentioned in another comment these styles are handled by the popup now and I just failed to remove the unused file

tomdye commented on src/context-menu/context-menu.m.css:
```diff
@@ -0,0 +1,13 @@
+	z-index: 11;
+}
+
+.underlay {
```
is this still needed? I'd have imagined that the popup deals with the underlay?

maier49 replied:
Removed

tomdye commented on src/context-menu/context-menu.m.css:
```diff
@@ -0,0 +1,13 @@
+.root {
+	position: absolute;
+	z-index: 11;
```
we need to make these z-indexes higher really, i believe we have some guidance on this somewhere

tomdye replied:
https://github.com/dojo/widgets#z-index-layering

maier49 replied:
I've removed these CSS files, as you mentioned in another comment these styles are handled by the popup now and I just failed to remove the unused file

tomdye commented:
@maier49 btw i updated this PR with latest widgets to benefit from no body scroll when popup is open

dojo-site commented:
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-jqlpubhk8.now.sh
* Tests Deployment: https://dojowidgets-lemr5azf0.now.sh
